### PR TITLE
initial implementation for funding service vault

### PR DESCRIPTION
**Pull request by 0xwalid**
Basically a wallet for the funding protocol. the idea is to keep it as simple and secure as possible and separate the business logic from the funds storage.
ERC20 support is pending so i didn't write tests for it

### CODE REVIEW

File: smart-contracts/ethereum/contracts/FundingVault.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
+pragma solidity ^0.4.24;
+
+
+// TODO move ERC20 interface to a separate file
+
+contract ERC20Basic {
```
Comment (JordoBrooks) [inline smart-contracts/ethereum/contracts/FundingVault.sol]: Should we be using the "interface" keyword here? (http://solidity.readthedocs.io/en/v0.4.21/contracts.html#interfaces)

**Comment by ericmuyser** [comment]
Introducing Chai...

Argument for Chai:
- More expressive
- Automatic exception catching
- Thought is it may be more common in the future

Argument against Chai:
- It may be additional work to become familiar with the syntax for devs coming from normal assertions / we may receive non-chai tests in PRs
- It may be less clear for non-native English speakers

Examples in the wild:
- Zepplin
  - Could not find any negative sentiment in Issues or PRs
  - https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/test/payment/SplitPayment.test.js
  - https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/test/token/ERC721/ERC721MintBurn.behaviour.js
- POA Network
  - https://github.com/poanetwork/poa-network-consensus-contracts/blob/master/test/voting_to_change_keys_test.js

File: package.json [hunk 2]
```diff
@@ -24,6 +24,9 @@
   },
   "homepage": "https://github.com/hyperbridge/funding-protocol#readme",
   "dependencies": {
+    "chai": "^4.1.2",
+    "chai-as-promised": "^7.1.1",
+    "chai-bignumber": "^2.0.2",
```
Comment (ericmuyser) [inline package.json]: These should be devDependencies

File: smart-contracts/ethereum/contracts/FundingVault.sol [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    require(getBalance() >= _amount);
+
+    emit ETHWithdrawn(_receiver, _amount);
+
+    _receiver.transfer(_amount);
+  }
+
+  event ERC20Withdrawn(address ERC20, address receiver, uint amount);
+
+  function withdrawERC20(address _ERC20Adress, uint _amount, address _receiver) public fundingServiceOnly activeOnly {
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/FundingVault.sol]: `_ERC20Adress` typo - `_tokenAddress`?

File: smart-contracts/ethereum/contracts/FundingVault.sol [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      require(_amount > 0);
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/FundingVault.sol]: We wan't to check the receiver isn't 0x0 here too?

File: smart-contracts/ethereum/contracts/FundingVault.sol [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    fundingService = _fundingService;
+  }
+
+  function setActive(bool val) public ownerOnly {
+    active = val;
+  }
+
+  event ETHDeposited(uint amount);
+
+  function depositeETH() public payable  fundingServiceOnly activeOnly {
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/FundingVault.sol]: `deposit` not `deposite`

File: smart-contracts/ethereum/contracts/FundingVault.sol [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  modifier activeOnly() {
+    require(active == true);
+    _;
+  }
+
+  address fundingService;
+
+  address owner;
+
+  //TODO import pausable?
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/FundingVault.sol]: Good idea

File: smart-contracts/ethereum/contracts/FundingVault.sol [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    _receiver.transfer(_amount);
+  }
+
+  event ERC20Withdrawn(address ERC20, address receiver, uint amount);
+
+  function withdrawERC20(address _ERC20Adress, uint _amount, address _receiver) public fundingServiceOnly activeOnly {
+      require(_amount > 0);
+      ERC20 token = ERC20(_ERC20Adress);
+      require (token.balanceOf(this) > _amount);
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/FundingVault.sol]: Not `>=`?

**Comment by ericmuyser** [comment]
Can we convert to 4 spaces for tabs btw? (consistency)
I'll add a .editorconfig

File: smart-contracts/ethereum/contracts/Ownable.sol [hunk 8]
```diff
@@ -0,0 +1,64 @@
+pragma solidity ^0.4.24;
+
+
+/**
+ * @title Ownable
+ * @dev The Ownable contract has an owner address, and provides basic authorization control
+ * functions, this simplifies the implementation of "user permissions".
+ */
+contract Ownable {
```
Comment (JordoBrooks) [inline smart-contracts/ethereum/contracts/Ownable.sol]: This has now already been added in master, and can be imported from "./openzeppelin/Ownable.sol"

**Comment by ericmuyser** [comment]
LGTM 👍

**Pull request merged**